value/log.go

```diff
@@ -11,7 +11,25 @@ func logn(c Context, v Value) Value {
 }
 
 func logBaseU(c Context, u, v Value) Value {
-	return c.EvalBinary(logn(c, v), "/", logn(c, u))
+	// Handle the integer part exactly when the arguments are exact.
+	var i Int
+	switch u.(type) {
+	case Int, BigInt, BigRat:
+		if toBool(c.EvalBinary(u, ">", Int(1))) {
+			for toBool(c.EvalBinary(v, ">=", u)) && toBool(c.EvalBinary(c.EvalBinary(v, "mod", u), "==", Int(0))) {
+				i++
+				v = c.EvalBinary(v, "/", u)
+			}
+			if toBool(c.EvalBinary(v, "==", Int(1))) {
+				return i
+			}
+		}
+	}
+	f := c.EvalBinary(logn(c, v), "/", logn(c, u))
+	if i != 0 {
+		f = c.EvalBinary(f, "+", i)
+	}
+	return f
```

Comment:
I was confused by the CL description. I was expecting a call to big.Int to handle the fact (not usually worried about elsewhere) that a transcendental operation can become integral. You're doing something else here, which is to use integer arithmetic to evaluate it. Also it's pretty inefficient to go through EvalBinary for such simple checks like > and >=. I might put methods or functions into the implementations of the three types.

For the topic actually mentioned in the CL, it might be OK to put a shrink call into the transcendental operations, but I'm of two minds.

Reply:
The shrink doesn't work for the ln-based approach, because the converged ln 2 and ln 8 round in different directions, so that dividing one by the other will never be exactly 3.

I've pushed an update that computes the log one bit at a time, instead of counting up to it.